### PR TITLE
Move a large chunk of the setup.py configuration to setup.cfg

### PR DESCRIPTION
**Pull request by astrofrog**
As of setuptools 30 (released December 2016), setuptools can use options and metadata directly from ``setup.cfg``:

https://setuptools.readthedocs.io/en/latest/setuptools.html#configuring-setup-using-setup-cfg-files

Note that Python 3.6 was released the same month, so in effect any Python 3.6 and 3.7 users should definitely have recent enough versions of setuptools, and most Python 3.5 users will likely also have a recent version, so I think this is probably relatively safe. By the time of the 3.2 release, setuptools 30 will be almost 2.5 years old.

*(this is part of a series of PRs to simplify setup.py)*

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -35,6 +80,3 @@ exclude = extern,*parsetab.py,*lextab.py
 
 [pycodestyle]
 exclude = extern,*parsetab.py,*lextab.py
-
-[metadata]
-license_file = LICENSE.rst
```
Comment (pllim) [inline setup.cfg]: Don't need `license_file` anymore?

File: setup.cfg [hunk 2]
```diff
@@ -1,3 +1,48 @@
+[metadata]
+name = astropy
+version = 3.2.dev
+provides = astropy
+author = The Astropy Developers
+author_email = astropy.team@gmail.com
+license = BSD 3-Clause License
+url = http://astropy.org
+description = Community-developed python astronomy tools
+keywords = astronomy, astrophysics, cosmology, space, science,units, table, wcs, samp, coordinate, fits, modeling, models, fitting, ascii],
```
Comment (pllim) [inline setup.cfg]: Mismatched `]`?

**Comment by bsipocz** [comment]
The change in the minimum required setuptools version probably deserves a mention in the changelog.

File: setup.py [hunk 3]
```diff
@@ -3,12 +3,25 @@
...
+    sys.exit(1)
+
+from setuptools.config import read_configuration
+conf = read_configuration('setup.cfg')
+
 # This is the same check as astropy/__init__.py but this one has to
 # happen before importing ah_bootstrap
-__minimum_python_version__ = '3.5'
-if sys.version_info < tuple((int(val) for val in __minimum_python_version__.split('.'))):
+minimum_python_version = conf['options']['python_requires'].replace('>', '').replace('<', '').replace('=', '')
```
Comment (mhvk) [inline setup.py]: Why replace the `<` - that should error, no?

**Comment by mhvk** [comment]
This looks very nice. Should ensure we propagate this to the affiliated package templates as well (echoing my https://github.com/astropy/astropy/pull/8248#issuecomment-445286612 about tox that we should aim to have astropy run as much as possible a setup that we provide in the template).

File: setup.py (hunk 3 above)
Comment (pllim) [reply setup.py]: Triple replace is messing with my head. What is this trying to do?

Comment (astrofrog) [reply setup.py]: Basically this is changing a requirement (from e.g. ``>=3.5``) to an actual minimum version (``3.5``). I agree ``<`` shouldn't really be used though.

File: setup.py [hunk 4]
```diff
@@ -3,12 +3,26 @@
 
 import sys
 
+from distutils.version import LooseVersion
+
+# We require setuptools 30.3.0 or later for the configuration in setup.cfg to
+# work properly.
+import setuptools
+if LooseVersion(setuptools.__version__) < LooseVersion('30.3.0'):
```
Comment (eteq) [inline setup.py]: I just looked and it seems like 30.3 was released Dec 6, 2016.  So >2 years, which is probably a pretty safe bet (even the LTS would be compatible, although presumably we're not doing that anyway).

**Comment by astrofrog** [comment]
This is not copied from the package template (which has its own PR for a similar approach here: astropy/package-template#375 - note that previously we parsed setup.cfg ourselves). But I do like the idea of trying to make sure our ``setup.py`` files are basically the same. I'll see what I can do in that respect.

**Comment by bsipocz** [comment]
> This looks good to me, but one change that's missing: the release documentation (docs/development/releasing.rst) should be updated to say that the version should be changed in setup.cfg, rather than setup.py as it currently says. (I think it's step 10? although it might appear a couple other places in that doc).

You can dump that to the list in https://github.com/astropy/astropy/issues/5120, at some point I will to the point to address that issue.

**Comment by Cadair** [comment]
I feel like getting to the point where you can retemplate astropy core with the package template would be a great test of the flexibility of the package template ;)

**Comment by bsipocz** [comment]
😱

**Comment by astrofrog** [comment]
One difference between astropy core and the package template is that ``astropy/__init__.py`` defines:

```
__minimum_python_version__ = '3.5'
__minimum_numpy_version__ = '1.13.0'
```

and then actually gets ``astropy.__minimum_numpy_version__`` to use in ``setup_requires``. We also have custom logic (in core) to add cython and jinja2 as setup_requires dependencies for non-release versions.

So at the moment it's not entirely feasible to have the same ``setup.py`` file, though I think some of these could probably be cleaned up - in particular if we specify the minimum numpy version in ``install_requires`` then importing astropy should fail if the numpy version is too old, which means we wouldn't need our custom logic.

**Comment by Cadair** [comment]
Those things all sound solveable to me. I will open an independent issue so we can attempt to keep track of it.

**Comment by eteq** [comment]
(just in case it got missed in the above discussion: the "request changes" I had in my review here was about the needed change in the release instructions)

**Comment by astrofrog** [comment]
@eteq - with a little delay, I've now addressed your comment about the release instructions.

**Comment by eteq** [comment]
LGTM now, and the appveyor is a red herring, so will go ahead and merge.

**Pull request merged**